File: apps/web/src/api/folders.ts
/**
 * Folders — the register's filing tree (#WS2). Every type comes from the
 * generated `@loft/ts-client` (pydantic → OpenAPI → TS; CLAUDE.md DRY rule);
 * nothing here re-describes a shape the server already documents.
 *
 * The four decisions this layer is a client of are stated on the server
 * (`py_kit/schemas/folders.py`) and are worth repeating only where they change
 * what a caller may DO:
 *
 * - a folder belongs to ONE drawer, so every call carries a `kind`;
 * - "unfiled" is `folder_id: null` and is a real destination, never an omission
 *   — which is why `moveDocument` takes `string | null` and never an optional;
 * - deleting a folder that still holds things is REFUSED, with the contents
 *   named (`FolderNotEmptyError`), the same grammar the document delete uses;
 * - the counts on a folder come from the server. Nothing in this file derives
 *   one, and the register must not either: at the root the browser is holding
 *   the *unfiled* documents, so any count it made itself would be about a
 *   different set.
 */
import type { components, GatewayClient } from "@loft/ts-client/gateway";

import { gatewayClient } from "./client";
import { envelopeCode, envelopeMessage } from "./envelope";

/** One folder as stored, with its two DIRECT (non-recursive) counts. */
export type FolderResponse = components["schemas"]["FolderResponse"];
/**
 * Which drawer a folder belongs to: `part` | `assembly` | `drawing`.
 * Read off the response's own field — a bare `Literal` alias does not become a
 * named OpenAPI component, so this is where the generated union actually lives.
 */
export type FolderKind = FolderResponse["kind"];
/** One thing inside a folder whose delete was refused (a sub-folder or a doc). */
export type FolderMember = components["schemas"]["FolderMember"];

/**
 * A folder delete was refused because the folder still holds things (409
 * `folder_not_empty`) — the filing-level sibling of
 * `DocumentHasDependentsError`.
 *
 * Typed for the same reason: the user's next action is to move those items out,
 * so the refusal has to name them. "This folder is not empty" would end the
 * conversation.
 */
export class FolderNotEmptyError extends Error {
  constructor(
    readonly contents: FolderMember[],
    message: string,
  ) {
    super(message);
    this.name = "FolderNotEmptyError";
  }
}

/**
 * Runtime narrowing of a 409 envelope's `details.contents`, or null.
 *
 * The SHAPE is the generated contract; this only checks that an `unknown` body
 * really is that shape before trusting it. Returns null rather than a partial
 * list — half a contents list is worse than none, because a user would move out
 * exactly what they were shown and hit the same refusal again.
 */
export function parseFolderContents(body: unknown): FolderMember[] | null {
  if (typeof body !== "object" || body === null) return null;
  const error = (body as Record<string, unknown>).error;
  if (typeof error !== "object" || error === null) return null;
  const details = (error as Record<string, unknown>).details;
  if (typeof details !== "object" || details === null) return null;
  const raw = (details as Record<string, unknown>).contents;
  if (!Array.isArray(raw)) return null;
  const contents: FolderMember[] = [];
  for (const entry of raw) {
    if (typeof entry !== "object" || entry === null) return null;
    const { id, name, kind } = entry as Record<string, unknown>;
    if (typeof id !== "string" || typeof name !== "string") return null;
    if (
      kind !== "folder" &&
      kind !== "part" &&
      kind !== "assembly" &&
      kind !== "drawing"
    ) {
      return null;
    }
    contents.push({ id, name, kind });
  }
  return contents.length > 0 ? contents : null;
}

/** The caller's whole folder tree for one drawer, name-ordered. */
export async function fetchFolders(
  kind: FolderKind,
  client: GatewayClient = gatewayClient,
): Promise<FolderResponse[]> {
  const { data, error } = await client.GET("/api/v1/folders", {
    params: { query: { kind } },
  });
  if (error !== undefined) {
    throw new Error(
      envelopeMessage(error, "Your folders could not be loaded."),
    );
  }
  return data.folders;
}

/** Create a folder at the root (`parentId` null) or inside another (201). */
export async function createFolder(
  name: string,
  kind: FolderKind,
  parentId: string | null = null,
  client: GatewayClient = gatewayClient,
): Promise<FolderResponse> {
  const { data, error } = await client.POST("/api/v1/folders", {
    body: { name, kind, parent_id: parentId },
  });
  if (error !== undefined) {
    if (envelopeCode(error) === "folder_name_taken") {
      throw new Error(
        envelopeMessage(error, `A folder named “${name}” already exists here.`),
      );
    }
    throw new Error(envelopeMessage(error, "The folder could not be created."));
  }
  return data;
}

/** Rename a folder. Renaming cannot move it — that is `moveFolder`. */
export async function renameFolder(
  folderId: string,
  name: string,
  client: GatewayClient = gatewayClient,
): Promise<FolderResponse> {
  const { data, error } = await client.PATCH("/api/v1/folders/{folder_id}", {
    params: { path: { folder_id: folderId } },
    body: { name },
  });
  if (error !== undefined) {
    throw new Error(envelopeMessage(error, "The folder could not be renamed."));
  }
  return data;
}

/**
 * Re-parent a folder; `parentId: null` moves it to the root of its drawer.
 *
 * A move into the folder itself or into one of its own descendants is refused
 * (422 `folder_cycle`) — the server decides that, because only the server holds
 * the whole tree. The message it sends is what the user sees.
 */
export async function moveFolder(
  folderId: string,
  parentId: string | null,
  client: GatewayClient = gatewayClient,
): Promise<FolderResponse> {
  const { data, error } = await client.POST(
    "/api/v1/folders/{folder_id}/move",
    {
      params: { path: { folder_id: folderId } },
      body: { parent_id: parentId },
    },
  );
  if (error !== undefined) {
    throw new Error(envelopeMessage(error, "The folder could not be moved."));
  }
  return data;
}

/**
 * Delete an EMPTY folder (204).
 *
 * A folder that still holds sub-folders or documents is refused with a 409 that
 * NAMES them — thrown as `FolderNotEmptyError` so the register can list them.
 * Never a cascade: deleting a folder cannot delete a document.
 */
export async function deleteFolder(
  folderId: string,
  client: GatewayClient = gatewayClient,
): Promise<void> {
  const { error } = await client.DELETE("/api/v1/folders/{folder_id}", {
    params: { path: { folder_id: folderId } },
  });
  if (error !== undefined) {
    const contents = parseFolderContents(error);
    if (contents !== null) {
      throw new FolderNotEmptyError(
        contents,
        envelopeMessage(error, "That folder still holds items."),
      );
    }
    throw new Error(envelopeMessage(error, "The folder could not be deleted."));
  }
}

/**
 * The ancestor chain of a folder, ROOT FIRST — the breadcrumb.
 *
 * Pure, over the flat tree the list endpoint returns, and defensive about a
 * chain that does not terminate: it stops at the number of folders that exist,
 * so a malformed tree renders a short breadcrumb rather than hanging the tab.
 * Returns an empty array for the root view.
 */
export function folderPath(
  folders: readonly FolderResponse[],
  folderId: string | null,
): FolderResponse[] {
  const byId = new Map(folders.map((folder) => [folder.id, folder]));
  const chain: FolderResponse[] = [];
  let current = folderId === null ? undefined : byId.get(folderId);
  while (current !== undefined && chain.length <= folders.length) {
    chain.unshift(current);
    current =
      current.parent_id === null ? undefined : byId.get(current.parent_id);
  }
  return chain;
}

/**
 * A folder's full path as text ("Gearbox / Housings"), for a row label and for
 * the move picker. `null` is the drawer root, which has a NAME rather than an
 * empty string, because "Unfiled" is a place a document can be.
 */
export function folderPathLabel(
  folders: readonly FolderResponse[],
  folderId: string | null,
): string {
  if (folderId === null) return UNFILED_LABEL;
  const chain = folderPath(folders, folderId);
  // A folder we were not given (deleted or filtered out) is reported as
  // unknown rather than silently drawn as the root: the register would
  // otherwise claim a document is unfiled when the server says it is not.
  return chain.length === 0
    ? "Unknown folder"
    : chain.map((folder) => folder.name).join(" / ");
}

/** What the root of a drawer is called wherever a folder name would go. */
export const UNFILED_LABEL = "Unfiled";

/** The folders directly inside `parentId` (null = the drawer's top level). */
export function childFolders(
  folders: readonly FolderResponse[],
  parentId: string | null,
): FolderResponse[] {
  return folders.filter((folder) => folder.parent_id === parentId);
}

File: packages/py-kit/src/py_kit/schemas/folders.py
"""Folder DTOs — the filing structure behind all three registers (#WS2).

#WS1 shipped search, sort, rename, duplicate and a delete that names its
dependents, and stopped at folders rather than draw a folder rail backed by
nothing. This module is the contract that backs it. Four decisions are made
here, in the open, because each one is a thing a surface could otherwise
quietly get wrong:

**1. A folder belongs to ONE document kind** (:data:`FolderKind`). The product's
information architecture is already three drawers — ``/`` parts, ``/assemblies``,
``/drawings`` — each with its own register, its own count and its own filter. A
single tree spanning all three would put folders in the parts register that hold
no parts (chrome that measures nothing, mandate 3a) and would force the count
readout to either lie about what is in the drawer or report a number the drawer
cannot show. A shared tree is the right model only alongside a single unified
register, which is a different, larger change; inventing it under three drawers
would produce exactly the empty-folder surface this repo keeps deleting.

**2. "Unfiled" is a real state, not a synthetic root** (``folder_id`` is
NULLABLE). Every document that exists today is unfiled and stays reachable with
no backfill, no per-owner root rows to mint, and no way for the root to be
missing. The register's root view IS the unfiled set plus the top-level folders,
so nothing is hidden by the model itself — and the drawer-wide filter (which
searches every folder and labels each hit with where it lives) is what
guarantees a filed document can always be found again.

**3. Names are unique PER FOLDER**, not per owner (see
``documents.db`` for the pair of partial unique indexes that enforce it,
including for the unfiled set — a plain composite UNIQUE would let two unfiled
"Bracket"s exist, because SQL treats NULLs as distinct). Two folders may each
hold a "Bracket"; one folder may not hold two.

**4. Deleting a non-empty folder is REFUSED** and the refusal names what is
inside (:class:`FolderContents`). The alternatives were cascade (deletes
documents the user never named — unacceptable for the one object in the product
that holds hours of work) and orphan-to-root (silently relocates work, so the
delete's real effect is invisible). Refusal is also the convention this repo
already runs on: it is the same 409-with-contents shape as the part delete's
409-with-dependents, so a user meets ONE refusal grammar, not two.

Deliberately NOT here: a recursive document count. Every count on the wire
(:attr:`FolderResponse.document_count`, :attr:`FolderResponse.child_folder_count`)
is DIRECT — what one level of the folder holds — because that is what one
GROUP BY can answer honestly and what the row a user is looking at is about. A
"47 parts" that silently included six sub-folders' worth would be the same
over-claim as a register health badge derived from fetch state.
"""

import uuid
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

#: Which drawer a folder lives in. A folder holds documents of its OWN kind
#: only; see the module docstring for why the tree is per-kind.
FolderKind = Literal["part", "assembly", "drawing"]

#: Upper bound for a folder name — shorter than a document's (200) on purpose:
#: a folder name is rendered inside breadcrumbs and inside row labels ("in
#: Gearbox"), where a 200-character name would wreck the line it sits in.
FOLDER_NAME_MAX_LENGTH = 120

#: Non-empty (post-strip), bounded folder name.
FolderName = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True, min_length=1, max_length=FOLDER_NAME_MAX_LENGTH
    ),
]

#: How deep the tree may go, counting the root level as 1. A bound exists so a
#: pathological tree cannot make the breadcrumb unrenderable or the ancestor
#: walk unbounded; 8 is far past any filing a person does by hand and is
#: reported as a plain 422 with the limit named, never a silent truncation.
MAX_FOLDER_DEPTH = 8


#: Description of the ``folder_id`` field every document response carries.
#: Written once (CLAUDE.md DRY) and referenced by ``PartResponse``,
#: ``AssemblyResponse`` and ``DrawingResponse`` — three copies of this sentence
#: would be three chances for one drawer to describe filing differently.
FOLDER_ID_DESCRIPTION = (
    "The folder this document is filed in, or null when it is UNFILED (at the "
    "root of its drawer). Null is a real state, not a missing value — see "
    "py_kit.schemas.folders. Changed only by the document's `/move` route, "
    "which is not a document edit: it moves neither the concurrency counter nor "
    "`updated_at`."
)


class FolderCreate(BaseModel):
    """Create a folder in one drawer, at the root or inside another folder."""

    name: FolderName = Field(
        description="Folder name; unique among its siblings (per parent, per "
        f"kind, per owner), whitespace-trimmed, 1-{FOLDER_NAME_MAX_LENGTH} chars"
    )
    kind: FolderKind = Field(
        description="Which drawer this folder belongs to. It may only ever hold "
        "documents of this kind (see module docstring)."
    )
    parent_id: uuid.UUID | None = Field(
        default=None,
        description="Containing folder, or null for a top-level folder. Must be "
        "the caller's own folder OF THE SAME KIND.",
    )


class FolderRename(BaseModel):
    """Rename a folder. Renaming cannot move it — that is :class:`FolderMove`.

    Separate verbs on purpose: a PATCH carrying an optional ``parent_id`` cannot
    tell "leave it where it is" (field omitted) from "move it to the root"
    (field null), and a move that silently did nothing because a client sent the
    wrong one of those is precisely the "reported success while the document is
    still in the old place" defect this slice is being held to.
    """

    name: FolderName = Field(description="New folder name")


class FolderMove(BaseModel):
    """Re-parent a folder. ``parent_id: null`` moves it to the root.

    Refused (422) when the target is the folder itself or one of its own
    descendants — that would detach the subtree from the register entirely, so
    it is a cycle in effect as well as in the graph.
    """

    parent_id: uuid.UUID | None = Field(
        description="The new containing folder, or null to move to the root of "
        "its drawer. Required — null is an explicit destination, not an omission."
    )


class DocumentMove(BaseModel):
    """File a document into a folder. ``folder_id: null`` un-files it.

    Applies to a part, an assembly or a drawing (one DTO — the verb is identical
    and only the route differs). Filing is NOT a document edit: it does not bump
    the document's concurrency counter and does not move ``updated_at``, so the
    register's LAST WORKED column keeps meaning "someone worked on it" rather
    than "someone tidied up". Same reasoning as the last-evaluate record write.
    """

    folder_id: uuid.UUID | None = Field(
        description="The destination folder — must be the caller's own and OF "
        "THIS DOCUMENT'S KIND — or null to leave the document unfiled at the "
        "root of its drawer. Required; null is an explicit destination."
    )


class FolderResponse(BaseModel):
    """A folder as stored, plus the two counts a register row may state.

    Both counts are DIRECT (this folder's own children), computed server-side in
    the list query. A register renders them; it never derives one from the rows
    it happens to be holding, because at the root it is holding only the unfiled
    documents and would count them as the folder's.
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    owner_id: uuid.UUID = Field(description="Owning user id (gateway-verified)")
    kind: FolderKind = Field(description="The drawer this folder belongs to")
    name: str
    parent_id: uuid.UUID | None = Field(
        description="Containing folder, or null for a top-level folder"
    )
    document_count: int = Field(
        ge=0,
        description="Documents filed DIRECTLY in this folder — not counting "
        "sub-folders' contents. A register may print this; it may not add it up.",
    )
    child_folder_count: int = Field(
        ge=0, description="Folders directly inside this one (not recursive)"
    )
    created_at: datetime
    updated_at: datetime


class FolderListResponse(BaseModel):
    """Every folder of ONE kind for the caller — the whole tree, one query.

    The whole tree rather than one level: the register needs ancestors for the
    breadcrumb and the full set for the move picker, and a per-level fetch would
    make "which folder is this document in?" an N+1. A wrapper leaves room for
    pagination the day a tree is big enough to need it — which, for a structure
    a human types by hand, it will not be.
    """

    folders: list[FolderResponse]


#: What can be inside a folder: sub-folders and documents of the folder's kind.
FolderMemberKind = Literal["folder", "part", "assembly", "drawing"]


class FolderMember(BaseModel):
    """One thing inside a folder whose delete was refused.

    Deliberately NOT reusing :class:`~py_kit.schemas.workspace.DocumentDependent`:
    that model means "something REFERENCES you" and its ``kind`` documents why a
    part can never appear in it. Membership is the opposite relation, and a part
    is its commonest member. Two relations, two models; one shared refusal
    GRAMMAR (409, ``details`` naming what blocks it), which is the part that
    matters to a user.
    """

    id: uuid.UUID
    name: str = Field(description="Its name, as filed — for the refusal message")
    kind: FolderMemberKind = Field(
        description="'folder' for a sub-folder, else the document kind"
    )


class FolderContents(BaseModel):
    """The ``details`` payload of a non-empty-folder 409 — everything inside.

    A LIST, never a count: the caller's next action is to move those things out,
    which needs their names. Sub-folders first, then documents, each
    alphabetical, so two identical refusals read identically.
    """

    contents: list[FolderMember] = Field(
        description="Direct children of the folder; never empty (an empty folder "
        "deletes cleanly)"
    )


class FolderNotEmptyError(BaseModel):
    """The ``error`` member of a non-empty-folder 409, with typed ``details``."""

    code: str = Field(description="'folder_not_empty' — the code a client branches on")
    message: str = Field(description="Human summary; the register shows the list")
    details: FolderContents
    request_id: str | None = Field(
        default=None, description="Correlation id of the refused request"
    )


class FolderNotEmptyEnvelope(BaseModel):
    """A non-empty-folder 409 as it appears on the wire (standard envelope).

    Declared as the documented 409 model of the folder delete route so it reaches
    ``packages/contracts`` and therefore the generated TS client — the register
    reads ``details.contents`` as a TYPE rather than narrowing an ``unknown``.
    """

    error: FolderNotEmptyError
